homeassistant/components/synology_dsm/backup.py

```diff
@@ -0,0 +1,217 @@
+"""Support for Synology DSM backup agents."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Callable, Coroutine
+import logging
+from typing import Any
+
+from synology_dsm.api.file_station import SynoFileStation
+from synology_dsm.exceptions import SynologyDSMAPIErrorException
+
+from homeassistant.components.backup import AgentBackup, BackupAgent, BackupAgentError
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.aiohttp_client import ChunkAsyncStreamIterator
+from homeassistant.helpers.json import json_dumps
+from homeassistant.util.json import JsonObjectType, json_loads_object
+
+from .const import (
+    CONF_BACKUP_PATH,
+    CONF_BACKUP_SHARE,
+    DOMAIN,
+    SYNOLOGY_DATA_BACKUP_AGENT_LISTENERS,
+)
+from .models import SynologyDSMData
+
+LOGGER = logging.getLogger(__name__)
+
+
+async def async_get_backup_agents(
+    hass: HomeAssistant,
+) -> list[BackupAgent]:
+    """Return a list of backup agents."""
+    if not (
+        entries := hass.config_entries.async_loaded_entries(DOMAIN)
+    ) or not hass.data.get(DOMAIN):
+        LOGGER.debug("No proper config entry found")
+        return []
+    agents: list[BackupAgent] = []
+    for entry in entries:
+        syno_data: SynologyDSMData = hass.data[DOMAIN][entry.unique_id]
```

Comment:
`HassKey` could be used here instead

Reply:
I think it would even be better to migrate to config entry runtime data?

Reply:
nevermind, I remember now, that I've skipped the migration to runtime data, because of the media_source platform :see_no_evil: 
Now we use `HassKey`, but we need to assert each time the `unique_id` of the config entry - the soution would be to switch to the `entry_id` but this would need much more efforts, since the `unique_id` is the serial number of the nas, so it is used as identifier in different places (_eq. media_source and detection of already configured config-entry_)

Reply:
https://github.com/home-assistant/core/pull/135227#discussion_r1927723922